### ISSUE TITLE
Pagers/Pollers must use a client's api-version

### ISSUE THREAD

**Issue by jhendrixMSFT** [Azure/azure-sdk-for-go#21408]
When creating a `Pager[T]` or `Poller[T]`, it must use the `api-version` of the client.

**Comment by lirenhe** [comment]
@jhendrixMSFT, could you share more about the design for this item?

As some of the mgmt. SDKs do not follow the guide of using same api-version in the polling and paging URLs, could you also include the flag to turn off this?  
similar to dotnet:  https://github.com/Azure/autorest.csharp/blob/fd461f8480978606928fa404b0ddc5a2c477b5c1/src/assets/Generator.Shared/NextLinkOperationImplementation.cs#L49

I also heard dotnet does not require paging to use client's api-version.

**Comment by jhendrixMSFT** [comment]
Can you please explain more about the mismatched api-version?  If you have an example service that would be great.

CC @JeffreyRichter

**Comment by JeffreyRichter** [comment]
The purpose of the api-version query parameter is for the CLIENT to tell the SERVICE the behavior that it expects. And this depends on what the client actually expects which depends on the api-version it thinks it's using which is encoded into the SDK package VERSION they are using. So, a service returning a URL with an api-version value and forcing a client to use that api-version value is fundamentally broken. 

For example, service-v2 defines a new property that does not exist in service-v1. If a client sending api-version=v2 (perhaps via a newer SDK package version) talks to the service but the SDK uses a URL with api-version=v1, the service will not return the new property. But the customer code "thinks" it is using v1 and expects the property to be set and the property will exist in the SDK structure. So now, we have data corruption.

Most of the time, a client version initiates a LRO/POLLER and continues polling/paging so the client uses the same version all the time and this is why things have worked pretty good all this time. However (especially for paging) this is not guaranteed to be the case. For example, an end-user is using a web browser which talks to an Azure-customer service and this service gets the 1st page of a collection and returns an HTML page back to the browser. Then the end-user wants to advance to the next page (sending the nextLink URL) which sends a request to the service to get the next page. But the service is behind a load balancer and the end-user request can be dispatched to any one of a number of front-end nodes in a cluster. The request might get sent to a node running a different version of the SDK package which may not understand (be able to deserialize) the response if it just uses the nextLink URL without changing the api-version.

If some (ARM) services today return an api-version other than what the CLIENT originally used, then this is a bug in the service as it makes no sense for the SERVICE to MANDATE the api-version that the client should use. That defeats the whole purpose of the api-version query parameter in the first place.

**Comment by serbrech** [comment]
Are we saying that if I `PUT MyResource` on version **2**, then the polling url will get a `api-version` querystring of **2**?
Or are we saying that if the client has a definition for the operationResult at v=1 then that's the version to use for polling that resource no matter what?

example:

**ManagedCluster** resource has api-version 2023-08-01
**OperationResult** resource has api-version 2020-01-01 - it has been stable for a long time.

ManagedCluster response returns an OperationResult url in the header.

1. Should the SDK detect that this is an operationResult resource and the api-version it knows about for it (i.e: 2020-01-01)?
2. Should the SDK expect the OperationResult to support 2023-08-01, and use this version for polling, because the ManagedCluster was created at this version?
3. Should the polling just be dumb, and poll what it's given? (Opaque url)
---
1. Is a reasonable behavior, but unnecessarily strict IMO. What if I give you a polling url that is not part of the Client? The poller is currently not unmarshalling to a concrete type anyways, it relies on http status code and provisioning status value.
2. I think this is wrong, it breaks the point of versioning per resource type, so assuming you're going for 1.
3. This is what happens today, why is it a problem? are you trying to reduce the switch case of the poller decision making?

IMO this would be a fine compromise that does not require a boolean:

- if the server gives you an api-version, you should use that (3.) - preferred but should be documented in the guideline.
- if it doesn't, you should do (1.) - use the version known by the client.

Overriding the value _explicitely_ sent by the service can only lead to some unexpected behavior and issues for the service team.

If you expect the polling URL to be a resource that the client knows about, then why are we sending the URL, and not a resourceID?

**Comment by JeffreyRichter** [comment]
Azure should NOT be versioning resource types; Azure versions services and all operations of a service version uniformly together. In fact, this is the definition of a service: operations that version uniformly together and then Azure can produce SDKs that target a service version.

So, a service operation has a version that the client uses to PUT to create a resource. The PUT response includes a header with a URL that the client should use to poll the status of the resource creation. This url CAN DEFAULT to the same api-version used by the client with it called PUT. I can CAN because it would also be OK if the url had no api-version query parameter at all because the client MUST replace the api-version query parameter with the value the client desires (our SDKs should be doing this). In almost all cases, the api-version values will be identical but, in some cases, the api-version values will be different, and this still has to allow the clients to work correctly.

The polling URL should be an operation supported by the service. This must happen for at least documentation purposes and for the client to know the schema of the response in order to know when polling should stop. One app should be able to start the operation (using 1 api-version) and another app (possibly using another api-version) MUST be able to poll for completion. So, the operation must be a documented operation of the service. Customers may or may not use an SDK - nothing in the HTTP/REST design should require the use of an SDK. 

When a client polls, only the client knows how it will process the poll operation response and what properties it expects to see in this response. This is why the client MUST pass the api-version that it understands, and the service MUST honor this value when crafting the response. This is the whole purpose of the api-version query parameter. A service can never dictate or mandate what api-version a client must use - this defeats the whole purpose of the api-version query parameter.

**Comment by serbrech** [comment]
the manifest of azure services defines api-version per resourcetype, not per resource provider.
Is there any warning to a team that 2 resources within a resource provider are on 2 different versions? I'm not aware of one. Manifest linting certainly does not prevent us from doing this... so we can't expect this to be the rule.
I sdk providing a guideline along these lines?

My understanding of Azure APIs, until your message, is that a resource is an API, the resource provider is just a namespace.

for AKS, ManagedClusters is a resource, and exposes a set of operations that makes up the ManagedCluster API.
the SDK has a ManagedClusterClient, so 1 client per resource.

but there is only 1 package for the resource provider.

the operationPolling is shared for all these resources. so it cannot be aligned with all resources. that means all resources within a resource provider need to be versioned together. it's a constraint that I had never heard of, and isn't respected my any RP AFAIK.

**Comment by JeffreyRichter** [comment]
Yes, I'm aware that ARM manifest files version resource types But, this has proven to have been a mistake that we'd like to improve upon now. Look at it from the customer perspective: The customer has to know the api-version for each resource-type when working with it. This is MUCH more cumbersome than having a single api-version per service (which is already pretty cumbersome). Also, if a service mandates a newer api-version value, the client may not be able to handle it. Similarly, if the service mandates an older api-version, the client may also not be able to handle it if there was a breaking change applied between versions of the resource type.

When ARM originally thought of versioning resource types, the end-to-end scenarios were not fully thought through, and we now see many problems with the original design and so we'd like to fix them. 

Furthermore, the notion of versioning resource types doesn't really make sense (especially to customers). For example, the api-version is passed as a query parameter to an "operation" which may or may not operate on a "resource" and so the api-version value is logically applied to the operation, not the resource. In addition, resources themselves (which can be created by a client using V2 but can be returned to a client using V1) means that the resources themselves MUST ALWAYS be both forward and backward compatible. But, the projection of that resource can be a V1 projection or a V2 projection and the projection is applied by the "operation" version. So again, it only makes sense that the api-version applies to the service operation and not to a resource type. In other words, resource types cannot be versioned and only exposed to customers who speak that specific version.

Note that I'm ignoring SDKs in the conversation above because Azure has many customers not using SDKs at all. But, if you add SDKs to the conversation, then customers need to pick an SDK that corresponds to the "service version" so they know what behavior and resource properties to expect. Having SDKs with different clients corresponding to different resource types complicates things substantially for customers. It is MUCH better for an SDK package version to map 1:1 with a service version and we are here for all data-plane services and for almost all RPs as well.

Fortunately, there are few RPs that have this issue so it shouldn't be too difficult to correct/improve as we go forward. Perhaps @rkmanda can assist with getting a linting rule applied and we can reach out to the RP teams that are currently violating the "new & improved" thinking.

Having a single polling operation for all resource types sounds like a leaked abstraction to me. The ONLY way to make this work is if the polling operation never versions or introduces a breaking change. If it does, then polling just breaks for ALL clients.

**Comment by lirenhe** [comment]
> Can you please explain more about the mismatched api-version? If you have an example service that would be great.
> 
> CC @JeffreyRichter

Here is a real case in mgmt. SDK:  

https://github.com/Azure/azure-sdk-for-net/issues/34796#issuecomment-1461259256

And the source code:  https://github.com/Azure/azure-rest-api-specs/blob/644e559292ea3314c1636285f138b8696536b006/specification/resources/resource-manager/Microsoft.Resources/stable/2022-09-01/resources.json#L4334

The service responds with a 202 and Azure_AsyncOperation header that points to a different RP with a different api-version.
This is not common, but it does happen in mgmt. SDK.

**Comment by serbrech** [comment]
@JeffreyRichter thank you for the details. I appreciate the challenge. I think I'm mostly aligned with you. My main concern is the gap in our tooling and manifests that allows, or even encourages to break this guidance.

If I recall correctly AKS has done some work to align all our resources versions in the past year.

If this is enforced we still have some work to do, because we chose version separately fleet subfolder from AKS subfolder now. But that's special and would just push us to prioritize something we already have on our backlog (separate fleet operations)

What's still bothering me is that we return a URL. If the url is not opaque and gets tampered with by the client, returning an id would be simpler and clearer? Just thinking out loud.

@FumingZhang might be interested in this discussion

**Comment by JeffreyRichter** [comment]
I had some meetings with the container fleet people many months ago and we agrred that fleet is a separate service which can have its own api-version value that changes independantly from the non-fleet service.
We consider the nextLink URL opaque EXCEPT for the api-version query parameter. In affect, the rest of the URL is an "id". Would some kind of "continuationToken" id have been better? Probably yes, but its too late now to "fix" this. So, we just stick with the opaque URL and made the api-version a non-opaque piece of it.

**Comment by serbrech** [comment]
Q: Is this already enforce at the ARM level?

**Comment by serbrech** [comment]
Is there a documented guideline for Azure APIs that all resources within a resource provider namespace should be bumped together? (i.e: that operationsStatus API should be bumped together with all other resources).

**Comment by JeffreyRichter** [comment]
I do not think that such a guideline is documented but we should have this. Perhaps @rkmanda can help with this.
The idea is that a customer thinks in terms of Azure services & not in terms of Azure resource types. Therefore, a SERVICE has a version and all operations of the service expect the same api-version value. Azure SDKs have a 1+client per service and the api-version value is baked into a client and is used for ALL operations made to the service via that client. These are the reasons why this guideline is mandatory.
